src/services/navigationBar.ts

```diff
@@ -2,856 +2,551 @@
+                        return hasSomeImportantChild(item);
+                }
             }
-
-            // Return a module of either the given text in the first argument, or of the source file path
-            let defaultName = node.getSourceFile().fileName;
-            if (callExpr.arguments[0].kind === SyntaxKind.StringLiteral) {
-                defaultName = ((callExpr.arguments[0]) as StringLiteral).text;
+            function hasSomeImportantChild(item: NavNode) {
+                return forEach(item.children, child => {
+                    const childKind = navKind(child);
+                    return childKind !== SyntaxKind.VariableDeclaration && childKind !== SyntaxKind.BindingElement;
+                });
             }
-            return getNavBarItem(defaultName, ScriptElementKind.moduleElement, [getNodeSpan(node.parent)]);
         }
+    }
 
-        function getFunctionOrClassExpressionItem(node: Node): NavigationBarItem {
-            if (node.kind !== SyntaxKind.FunctionExpression &&
-                    node.kind !== SyntaxKind.ArrowFunction &&
-                    node.kind !== SyntaxKind.ClassExpression) {
-                return undefined;
-            }
+    function convertToTopLevelItem(n: NavNode): NavigationBarItem {
+        const spans = [getNodeSpan(n.node)];
+        return {
+            text: getItemName(n.node),
+            kind: nodeKind(n.node),
+            kindModifiers: navModifiers(n),
+            spans,
+            childItems: map(n.children, convertToChildItem),
+            indent: n.indent,
+            bolded: false,
+            grayed: false
+        };
+
+        function convertToChildItem(n: NavNode): NavigationBarItem {
+            const nodes = [n.node];
+            if (n.additionalNodes) {
+                nodes.push(...n.additionalNodes);
+            }
+            const spans = map(nodes, getNodeSpan);
+            return {
+                text: getItemName(n.node),
+                kind: nodeKind(n.node),
+                kindModifiers: navModifiers(n),
+                spans,
+                childItems: [],
+                indent: 0,
+                bolded: false,
+                grayed: false
+            };
+        }
+    }
 
-            const fnExpr = node as FunctionExpression | ArrowFunction | ClassExpression;
-            let fnName: string;
-            if (fnExpr.name && getFullWidth(fnExpr.name) > 0) {
-                // The expression has an identifier, so use that as the name
-                fnName = declarationNameToString(fnExpr.name);
-            }
-            else {
-                // See if it is a var initializer. If so, use the var name.
-                if (fnExpr.parent.kind === SyntaxKind.VariableDeclaration) {
-                    fnName = declarationNameToString((fnExpr.parent as VariableDeclaration).name);
+    // TODO: We should just use getNodeKind. No reason why navigationBar and navigateTo should have different behaviors.

```

Comment:
This would be a big change though, either to navigationBar tests or navigateTo tests. And there would have to be discussion of which ScriptElementKinds are better. I would rather file an issue now and deal with it after this is committed. OK?